compiler/rustc_mir_transform/src/jump_threading.rs

```diff
@@ -0,0 +1,727 @@
+            if cost.cost() > MAX_COST {
+                return;
+            }
+
+            // Attempt to turn the `current_condition` on `lhs` into a condition on another place.
+            self.process_statement(bb, stmt, &mut state);
+
+            // When a statement mutates a place, assignments to that place that happen
+            // above the mutation cannot fulfill a condition.
+            //   _1 = 5 // Whatever happens here, it won't change the result of a `SwitchInt`.
+            //   _1 = 6
+            if let Some((lhs, tail)) = self.mutated_statement(stmt) {
+                state.flood_with_tail_elem(lhs.as_ref(), tail, self.map, ConditionSet::default());
+            }
+        }
+
+        if self.is_empty(&state) || depth >= MAX_BACKTRACK {
+            return;
+        }
+
+        let last_non_rec = self.opportunities.len();
+
+        let predecessors = &self.body.basic_blocks.predecessors()[bb];
+        if let &[pred] = &predecessors[..] && bb != START_BLOCK {
+            let term = self.body.basic_blocks[pred].terminator();
+            match term.kind {
+                TerminatorKind::SwitchInt { ref discr, ref targets } => {
+                    self.process_switch_int(discr, targets, bb, &mut state);
+                    self.find_opportunity(pred, state, cost, depth + 1);
+                }
+                _ => self.recurse_through_terminator(pred, &state, &cost, depth),
+            }
+        } else {
+            for &pred in predecessors {
+                self.recurse_through_terminator(pred, &state, &cost, depth);
+            }
+        }
+
+        let new_tos = &mut self.opportunities[last_non_rec..];
+        debug!(?new_tos);
+
+        // Try to deduplicate threading opportunities.
+        if new_tos.len() > 1
+            && new_tos.len() == predecessors.len()
+            && predecessors
+                .iter()
+                .zip(new_tos.iter())
+                .all(|(&pred, to)| to.chain == &[pred] && to.target == new_tos[0].target)
+        {
+            // All predecessors have a threading opportunity, and they all point to the same block.
+            debug!(?new_tos, "dedup");
+            let first = &mut new_tos[0];
+            *first = ThreadingOpportunity { chain: vec![bb], target: first.target };
+            self.opportunities.truncate(last_non_rec + 1);
+            return;
+        }
+
+        for op in self.opportunities[last_non_rec..].iter_mut() {
+            op.chain.push(bb);
+        }
+    }
+
+    /// Extract the mutated place from a statement.
+    #[instrument(level = "trace", skip(self), ret)]
+    fn mutated_statement(
+        &self,
+        stmt: &Statement<'tcx>,
+    ) -> Option<(Place<'tcx>, Option<TrackElem>)> {
+        match stmt.kind {
+            StatementKind::Assign(box (place, _))
+            | StatementKind::Intrinsic(box NonDivergingIntrinsic::Assume(
+                Operand::Copy(place) | Operand::Move(place),
```

Comment:
Hmm.

Question: I can understand how a `Move` operand would be interpreted as a mutation of the place being moved, but why is a `Copy` of a place being interpreted as a mutation of the place being copied?

Reply:
Although now that I think about it, I'm no longer sure that the place being moved should be treated as a mutation for the purposes of what `mutated_place` is doing...

if I understand correctly, the purpose of `mutated_place` is to cover cases like this during the backward walk over a block:

```
 _1 = 5 // Whatever happens here, it won't change the result of a `SwitchInt` later on
 _1 = 6 // because the SwitchInt will always be based on this assignment to `_1` 
```

So what you care about are mutations that definitively assign a fresh value to a place, not mutations that might or might not overwrite previous values.

Do I have that right?

----

In any case, it would be useful for you to spell out in its doc comment whether `mutated_place` is meant to produce an overapproximation ("these are all the places that *might* be mutated") or an underapproximation ("this is a list of places all of which are *guaranteed* to be initialized by this statement). Having that in my mental model would have made it easier for me to understand the rest of this code.

Reply:
> So what you care about are mutations that definitively assign a fresh value to a place, not mutations that might or might not overwrite previous values.
> 
> Do I have that right?

(It seems I did *not* have that right. 😄 )